toolchain/check/merge.cpp

```diff
@@ -167,4 +167,161 @@ auto ReplacePrevInstForMerge(Context& context, SemIR::NameScopeId scope_id,
+                              SemIR::InstBlockId new_param_refs_id,
+                              SemIR::InstId prev_decl_id,
+                              SemIR::InstBlockId prev_param_refs_id,
+                              llvm::StringLiteral param_diag_label,
+                              Substitutions substitutions) -> bool {
+  // This will often occur for empty params.
+  if (new_param_refs_id == prev_param_refs_id) {
+    return true;
+  }
+
+  // If exactly one of the parameter lists was present, they differ.
+  if (new_param_refs_id.is_valid() != prev_param_refs_id.is_valid()) {
+    CARBON_DIAGNOSTIC(RedeclParamListDiffers, Error,
+                      "Redeclaration differs because of {1}{0}parameter list.",
+                      llvm::StringLiteral, llvm::StringLiteral);
+    CARBON_DIAGNOSTIC(RedeclParamListPrevious, Note,
+                      "Previously declared with{1} {0}parameter list.",
+                      llvm::StringLiteral, llvm::StringLiteral);
+    context.emitter()
+        .Build(
+            new_decl_id, RedeclParamListDiffers, param_diag_label,
+            new_param_refs_id.is_valid() ? llvm::StringLiteral("") : "missing ")
+        .Note(prev_decl_id, RedeclParamListPrevious, param_diag_label,
+              prev_param_refs_id.is_valid() ? llvm::StringLiteral("") : "out")
+        .Emit();
+    return false;
+  }
+
+  CARBON_CHECK(new_param_refs_id.is_valid() && prev_param_refs_id.is_valid());
+  const auto new_param_ref_ids = context.inst_blocks().Get(new_param_refs_id);
+  const auto prev_param_ref_ids = context.inst_blocks().Get(prev_param_refs_id);
+  if (new_param_ref_ids.size() != prev_param_ref_ids.size()) {
+    CARBON_DIAGNOSTIC(
+        RedeclParamCountDiffers, Error,
+        "Redeclaration differs because of {0}parameter count of {1}.",
+        llvm::StringLiteral, int32_t);
+    CARBON_DIAGNOSTIC(RedeclParamCountPrevious, Note,
+                      "Previously declared with {0}parameter count of {1}.",
+                      llvm::StringLiteral, int32_t);
+    context.emitter()
+        .Build(new_decl_id, RedeclParamCountDiffers, param_diag_label,
+               new_param_ref_ids.size())
+        .Note(prev_decl_id, RedeclParamCountPrevious, param_diag_label,
+              prev_param_ref_ids.size())
+        .Emit();
+    return false;
+  }
+  for (auto [index, new_param_ref_id, prev_param_ref_id] :
+       llvm::enumerate(new_param_ref_ids, prev_param_ref_ids)) {
+    if (!CheckRedeclParam(context, param_diag_label, index, new_param_ref_id,
+                          prev_param_ref_id, substitutions)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+auto CheckRedeclParamsMatch(Context& context, const EntityInfo& new_entity,
+                            const EntityInfo& prev_entity,
+                            Substitutions substitutions) -> bool {
+  if (EntityHasParamError(context, new_entity) ||
+      EntityHasParamError(context, prev_entity)) {
+    return false;
+  }
```

Comment:
Is it worth trying to treat as a match if *both* have errors, and in the same position?

Alternately, would it be worth trying to only check for an error after a mismatch is found, right before emitting? i.e., so that the cost of error checking only happens on the error path. This is fetching the underlying param structs which may remain in cache, but even then it feels a bit like extra logic to we might be able to avoid.

Reply:
Might be interesting to pursue that, yeah. For now this is just moved from function merging without changing the logic, so I'm not going to handle this in this PR.